File: app/src/main/java/dev/bee/kanjianki/MainActivityStats.java
package dev.bee.kanjianki;

import dev.bee.kanjianki.core.RecordsBase;

import dev.bee.kanjianki.core.KanjiImpactAnalyzer;
import dev.bee.kanjianki.core.StatsTextCopy;
import dev.bee.kanjianki.data.StudyStatsStore;

import java.util.ArrayList;
import java.util.List;

abstract class MainActivityStats extends MainActivityGames {
    void renderStats() {
        base("stats");
        content.addView(MainActivityStatsCompose.statsScreenView(this));
    }

    List<KanjiImpactAnalyzer.Row> notHelpingRows(KanjiImpactAnalyzer.Report report) {
        return KanjiImpactAnalyzer.notHelpingRows(report);
    }

    List<String> ladderDistributionRows(StudyStatsStore.LadderHealthMetric metric) {
        List<String> rows = new ArrayList<>();
        for (RecordsBase.LadderRung rung : RecordsBase.LadderRung.values()) {
            rows.add(StatsTextCopy.ladderDistributionRow(rung, metric.countFor(rung)));
        }
        return rows;
    }

    List<String> weaknessImprovementExamples(StudyStatsStore.WeakKanjiImprovedMetric metric) {
        List<String> examples = new ArrayList<>();
        int maxExamples = Math.min(3, metric.examples.size());
        for (int i = 0; i < maxExamples; i++) {
            StudyStatsStore.KanjiImprovement example = metric.examples.get(i);
            examples.add(StatsTextCopy.weaknessImprovementExample(example.kanji, example.beforeWeakness, example.afterWeakness));
        }
        return examples;
    }

    List<String> supportGainExamples(StudyStatsStore.MatureSupportGainedMetric metric) {
        List<String> examples = new ArrayList<>();
        for (StudyStatsStore.KanjiSupportGain example : metric.examples) {
            examples.add(StatsTextCopy.supportGainExample(example.kanji, example.beforeMatureSupport, example.afterMatureSupport));
        }
        return examples;
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivityStats.kt
package dev.bee.kanjianki

import androidx.compose.foundation.layout.padding
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import dev.bee.kanjianki.core.HomeTextCopy
import dev.bee.kanjianki.core.KanjiImpactAnalyzer
import dev.bee.kanjianki.core.StatsTextCopy
import dev.bee.kanjianki.data.MatureSupportGainedSnapshot
import dev.bee.kanjianki.data.WeakKanjiImprovedSnapshot
import dev.bee.kanjianki.progress.progressAnalyticsSampleSnapshot
import dev.bee.kanjianki.progress.progressAnalyticsSnapshot
import kotlinx.coroutines.runBlocking

internal abstract class MainActivityStats : MainActivityGames() {
    override fun renderStats() {
        if (isScreenshotLaunchRequested()) {
            renderScreenshotStats()
            return
        }
        currentRoute = MainActivityBase.NAV_STATS_ROUTE
        currentHomeRouteRestoration = null
        loadRouteAsync(
            showLoading = {
                composeRoute(MainActivityBase.NAV_STATS_ROUTE) {
                    HomeRouteLoadingScreen(
                        title = HomeTextCopy.statsActionLabel(),
                        homeLabel = HomeTextCopy.homeLabel(),
                        onHome = ::renderHome,
                    )
                }
            },
            traceName = "stats-route",
            load = {
                val now = System.currentTimeMillis()
                val stats = runBlocking { statsUseCases.loadForDisplay(now) }
                val settings = runBlocking { settingsUseCases.load() }
                progressAnalyticsSnapshot(stats, now, settings.studyLadder)
            },
            render = { model ->
                composeRoute(
                    selected = MainActivityBase.NAV_STATS_ROUTE,
                ) {
                    ProgressAnalyticsDashboardScreen(state = model, onBrowseKanji = { renderBrowseKanji(it) })
                }
            },
        )
    }

    private fun renderScreenshotStats() {
        val state = progressAnalyticsSampleSnapshot(0L)
        composeRoute(
            MainActivityBase.NAV_STATS_ROUTE,
            initialScrollY = screenshotScrollY(),
            scrollPositionLabel = screenshotScrollPositionLabel(),
        ) {
            ProgressAnalyticsDashboardScreen(
                state = state,
                modifier = Modifier.padding(bottom = 24.dp),
            )
        }
    }

    fun notHelpingRows(report: KanjiImpactAnalyzer.Report?): List<KanjiImpactAnalyzer.Row> {
        return KanjiImpactAnalyzer.notHelpingRows(report)
    }

    fun weaknessImprovementExamples(metric: WeakKanjiImprovedSnapshot): List<String> {
        return metric.examples.take(3).map { example ->
            StatsTextCopy.weaknessImprovementExample(
                example.kanji,
                example.beforeWeakness,
                example.afterWeakness
            )
        }
    }

    fun supportGainExamples(metric: MatureSupportGainedSnapshot): List<String> {
        return metric.examples.map { example ->
            StatsTextCopy.supportGainExample(
                example.kanji,
                example.beforeMatureSupport,
                example.afterMatureSupport
            )
        }
    }
}
